sled-agent/src/rack_setup/service.rs

```diff
@@ -51,176 +93,450 @@ impl Service {
             );
-            let old_config: Config = toml::from_str(
-                &tokio::fs::read_to_string(&rss_config_path).await?,
-            )?;
-            if &old_config == config {
-                info!(
-                    self.log,
-                    "RSS config already applied from: {}",
-                    rss_config_path.to_string_lossy()
-                );
-                return Ok(());
-            }
-
-            // TODO(https://github.com/oxidecomputer/omicron/issues/724):
-            // We could potentially handle this case by deleting all
-            // datasets (in preparation for applying the new
-            // configuration), but at the moment it's an error.
-            warn!(
-                self.log,
-                "Rack Setup Service Config was already applied, but has changed.
-                This means that you may have datasets set up on this sled, but they
-                may not match the ones requested by the supplied configuration.\n
-                To re-initialize this sled:
-                   - Disable all Oxide services
-                   - Delete all datasets within the attached zpool
-                   - Delete the configuration file ({})
-                   - Restart the sled agent",
-                rss_config_path.to_string_lossy()
-            );
-            return Err(SetupServiceError::Configuration);
+            return Ok(());
         } else {
-            info!(
-                self.log,
-                "No RSS configuration found at {}",
-                rss_config_path.to_string_lossy()
-            );
+            info!(self.log, "RSS configuration has not been fully applied yet",);
         }
 
+        // Wait for either:
+        // - All the peers to re-load an old plan (if one exists)
+        // - Enough peers to create a new plan (if one does not exist)
+        let maybe_plan = self.load_plan().await?;
+        let expectation = if let Some(plan) = &maybe_plan {
+            PeerExpectation::Precise(plan.keys().map(|a| *a.ip()).collect())
+        } else {
+            PeerExpectation::Arbitrary(config.requests.len())
+        };
+        let addrs = self.wait_for_peers(expectation).await?;
+        info!(self.log, "Enough peers exist to enact RSS plan");
+
+        // If we created a plan, reuse it. Otherwise, create a new plan.
+        let plan = if let Some(plan) = maybe_plan {
+            info!(self.log, "Re-using existing allocation plan");
+            plan
+        } else {
+            info!(self.log, "Creating new allocation plan");
+            self.create_plan(config, addrs).await?
+        };
+
+        // NOTE: This is a "point-of-no-return" -- before sending any requests
+        // to neighboring sleds, ensure that we've recorded our plan to durable
+        // storage. This way, if the RSS power-cycles, it can idempotently
+        // execute the same allocation plan.
```

Comment:
the way this is worded seems like it's not done yet. did `create_plan` perform a flush to durable storage?